### PR TITLE
Adding prefetch mechanism to consumer

### PR DESCRIPTION
**Pull request by yoelsherwin**
Resolves memphisdev/memphis.py#766

### CODE REVIEW

**Comment by idanasulinmemphis** [comment]
Hi @yoelsherwin, this is amazing.
I think that something in the design is not good. You've created a single cache per connection, so if the user calls fetch_messages for more than one station, you will end up with mixed messages in the same cache.
Another issue is that you added a new option for the consumer function, max_cached_messages, and it is a bit confusing for the user since if you are creating a consumer and calling consumer.consume function, this feature has no impact.

What are your thoughts about these 2 points?

**Comment by yoelsherwin** [comment]
@idanasulinmemphis Regarding your first comment - you are absolutely correct. I'll fix this by changing the `cached_messages` type to a dictionary (from station name to array of cached messages).
Regarding your 2nd comment - I thought that the max cache size should be configured on consumer setup, rather than each time we call the `consume` / having a fixed cache size that is configurable. Having said that, I see your point. Do you prefer that it will be an optional parameter in the `consume` function?

**Comment by idanasulinmemphis** [comment]
> @idanasulinmemphis Regarding your first comment - you are absolutely correct. I'll fix this by changing the `cached_messages` type to a dictionary (from station name to array of cached messages). Regarding your 2nd comment - I thought that the max cache size should be configured on consumer setup, rather than each time we call the `consume` / having a fixed cache size that is configurable. Having said that, I see your point. Do you prefer that it will be an optional parameter in the `consume` function?

cached_messages should be a dict, its keys are the station names and its value is also a dict of consumers which it values are list of messages. Station messages are not necessarily needed for all consumers, for example - 1 consumer is in offset 1 in the station and another one has already reached offset 10

**Comment by yoelsherwin** [comment]
Great, I'll fix the cached_messages to be as you described.
Regarding the max cache size param - since it won't be passed when the consumer is being created, I'll add it as an optional param to the `fetch_messages` function. WDYT?

**Comment by idanasulinmemphis** [comment]
Maybe 2 times the batch size?
and I'd change the name of it to prefetch_cache_size

**Comment by yoelsherwin** [comment]
@idanasulinmemphis  Do you mean having a default value of batch_size * 2? Or not receiving this param at all and always using this as the number of msgs to load?
If you meant to have it as a default value, it looks like python can't rely on one param to define the default value of another param.
Regarding the name change - NP.

**Review by shay23b — COMMENTED** [review]
Hi @yoelsherwin, please note the comments and if you have any questions feel free to ask

**Comment by yoelsherwin** [comment]
Hi @shay23b , went over your comments and pushed a fix now. Would love to hear your thoughts :)

**Comment by shay23bra** [comment]
@yoelsherwin great! I will review ASAP, thank you

**Comment by shay23b** [comment]
Hey @yoelsherwin,
Will review soon, thanks for your effort!

**Comment by idanasulinmemphis** [comment]
Thanks for all the hard work @yoelsherwin, we will take it under our testing framework and will update

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Comment (idanasulinmemphis) [inline .idea/.gitignore:1]: please remove the .idea files from the PR

File: .idea/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+/shelf/
```
[the remark this answers is not in the document]
Comment (yoelsherwin) [reply .idea/.gitignore]: Added documentation in the readme file

File: .idea/.gitignore (hunk 1 above)
Comment (yoelsherwin) [reply .idea/.gitignore]: removed :)

File: README.md [hunk 3]
```diff
@@ -320,6 +320,7 @@ msgs = await memphis.fetch_messages(
   generate_random_suffix=False
   start_consume_from_sequence=1 # start consuming from a specific sequence. defaults to 1
   last_messages=-1 # consume the last N messages, defaults to -1 (all messages in the station))
+  max_cached_messages = 1000 # maximum size of messages in local cache, defaults to 1000
```
[the remark this answers is not in the document]
Comment (yoelsherwin) [reply README.md]: this file was already updated

File: memphis/memphis.py [hunk 4]
```diff
@@ -632,6 +640,7 @@ async def fetch_messages(
             generate_random_suffix (bool): false by default, if true concatenate a random suffix to consumer's name
             start_consume_from_sequence(int, optional): start consuming from a specific sequence. defaults to 1.
             last_messages: consume the last N messages, defaults to -1 (all messages in the station).
+            prefetch: false by default, if true then fetch messages from local cache (if exists) and load more messages into the cache.
```
[the remark this answers is not in the document]
Comment (yoelsherwin) [reply memphis/memphis.py:695]: This param was removed after @shay23b CR

**Comment by yoelsherwin** [comment]
Hi, @idanasulinmemphis , some of the changes you requested were outdated - they were removed after @shay23b CR. Some of them were relevant and fixed. Do you have any other comments after the changes I made? :)

File: memphis/memphis.py [hunk 5]
```diff
@@ -17,7 +17,14 @@
 import copy
 import json
 import ssl
+<<<<<<< HEAD
```
Comment (shay23b) [inline memphis/memphis.py:20]: There is a confilct here

File: memphis/memphis.py [hunk 6]
```diff
@@ -52,8 +59,8 @@ def __init__(self):
         self.station_schemaverse_to_dls = {}
         self.update_configurations_sub = {}
         self.configuration_tasks = {}
-        self.producers_map = {}
-        self.consumers_map = {}
+        self.producers_map = dict()
```
Comment (shay23b) [inline memphis/memphis.py:62]: not needed please return to what it was before

File: memphis/memphis.py [hunk 7]
```diff
@@ -52,8 +59,8 @@ def __init__(self):
...
+        self.consumers_map = dict()
```
Comment (shay23b) [inline memphis/memphis.py:63]: not needed please return to what it was before

**Review by shay23b — COMMENTED** [review]
Hey, @yoelsherwin awesome work, a few more changes and then we will run a few tests on the code before merging :)

**Comment by yoelsherwin** [comment]
Hi @shay23b , went over your comments and pushed an update :)

File: memphis/consumer.py [hunk 8]
```diff
@@ -156,6 +159,18 @@ async def main(/, host, username, password, station):
...
+        if prefetch and self.cached_messages:
+            if len(self.cached_messages) >= batch_size:
+                messages = self.cached_messages[:batch_size]
+                self.cached_messages = self.cached_messages[batch_size:]
+            else:
+                pulled_messages_size = len(self.cached_messages)
+                messages = self.cached_messages
+                self.cached_messages = []
+            self.load_messages_to_cache(batch_size * 2 - pulled_messages_size)
+            return messages
```
Comment (shay23b) [inline memphis/consumer.py:172]: This is called only if prefetch is true and the cached_messages length is bigger then 0, so this is never called, I don't see where this is initialized for the first time

**Review by shay23b — COMMENTED** [review]
Hi @yoelsherwin,
After running a few tests it seems it doesn't work,
I think this is the issue - the load_messages_to_cache is never called, I added there a comment

File: memphis/consumer.py (hunk 8 above)
Comment (yoelsherwin) [reply memphis/consumer.py:175]: You are correct, fixed the bug now

File: memphis/consumer.py [hunk 9]
```diff
@@ -156,6 +159,21 @@ async def main(host, username, password, station):
...
+                self.cached_messages = self.cached_messages[batch_size:]
+                number_of_messages_to_prefetch = batch_size * 2 - batch_size  # calculated for clarity
+            elif self.cached_messages:
+                number_of_messages_to_prefetch = batch_size * 2 - len(self.cached_messages)
+                messages = self.cached_messages
+                self.cached_messages = []
+            else:
+                number_of_messages_to_prefetch = batch_size * 2
+            self.load_messages_to_cache(number_of_messages_to_prefetch)
+            return messages
```
Comment (shay23b) [inline memphis/consumer.py:175]: If prefetch = true and self.cached_messages is empty then  messages is empty and being returned - instead should fetch messages to return and then load messages to cache

**Comment by yoelsherwin** [comment]
Hi @shay23b , sorry for the delay, was caught up in work. I'll debug it now and push a fix soon :)

**Comment by yoelsherwin** [comment]
Hi @shay23b , I fixed it now. sending you a description of how I tested it on Discord :) 
Also, one more question: ATM this is only supported as part of the `fetch_messages` and `fetch` functions. Should I add support to the `Consumer.consumer` function as well?

**Review by shay23b — APPROVED** [review]
Great work! Thank you for an amazing contribution! @yoelsherwin

**Review by shay23b — APPROVED** [review]

**Pull request merged**